Added Optical flow redo generation

reallybigname opened the pull request:
Option works in 2D/3D mode
Does a disposable generation before the actual generation to get the flow between previous image and the generation.  Then, it discards that generation and warps the init_sample based on the captured flow before the actual generation happens.
- Takes twice as long (obviously)
- It can introduce smoothing, especially over time, but also adds a type of artistic consistency between frames.
- can be mixed and matched with any other modes, optical flow cadence, video, etc...

Fixed issue where Image or Video color match could be overwitten at strength 0

reallybigname committed:
Added Optical flow redo generation

Option works in 2D/3D mode
Does a disposable generation before the actual generation to get the flow between previous image and the generation.  Then, it discards that generation and warps the init_sample based on the captured flow before the actual generation happens.
- Takes twice as long (obviously)
- It can introduce smoothing, especially over time, but also adds a type of artistic consistency between frames.
- can be mixed and matched with any other modes, optical flow cadence, video, etc...

reallybigname committed:
prevent color match on strength 0 if using Video or Image color coherence

Follow up to strength 0 color match reset... prevents color match sample being replaced when strength goes to 0 if using color coherence Video or color coherence Image.

reallybigname committed:
Fixed PIL version conflict

- replaced Image,Resampling.LANCZOS with PIL.Image.LANCZOS
- had to add import PIL
- had heard of people on colab encountering this issue. Should fix both hybrid video and image/video color matching.
- tested, works fine locally, so I see no reason not to change

reallybigname committed:
Hybrid Video Update

- Adds DIS Fine mode, and sets it as the new default
- Optical flow cadence and optical flow redo generation will now use DIS Fine, for pixel accuracy
- Adds flow guidance to hybrid flow. No setting needed, it just happens automatically, assisting the flow calculation by feeding in the last flow.
- Adds flow_factor_schedule for hybrid flow
- reduced some code for bordermode by setting a variable for the cv2 border mode translated from the user setting

reallybigname committed:
Fixed mistake

accidentally removed this crucial division! all better

hithereai committed:
Merge remote-tracking branch 'upstream/automatic1111-webui' into pr/474

hithereai commented:
default 2d without redo optical flow:

https://user-images.githubusercontent.com/121192995/225909785-f3965b20-9e3c-4d27-a40e-84ab01839c08.mp4


with redo:

https://user-images.githubusercontent.com/121192995/225909811-05e17ea9-37a1-4440-8d62-9ed72e17b8ae.mp4

hithereai committed:
Merge remote-tracking branch 'upstream/automatic1111-webui' into pr/474

reallybigname committed:
Added Diffusion Redo feature

Added slider for diffusion redo.
 - llterally does a diffusion multiple extra times, feeding it back into itself. It is suggested to turn steps down if you turn redo up.
- while redoing, seed is randomized and restored
- fixed Optical Flow Redo so that it also randomizes and restores the seed, making it less muddy.

reallybigname committed:
Merge branch 'automatic1111-webui' of https://github.com/reallybigname/deforum-for-automatic1111-webui into automatic1111-webui

kabachuha commented on scripts/deforum_helpers/render.py:
```diff
@@ -484,12 +485,26 @@ def render_animation(args, anim_args, video_args, parseq_args, loop_args, contro
         if anim_args.optical_flow_redo_generation and prev_img is not None and strength > 0:
             print("Optical Flow redo creating disposable diffusion before actual diffusion for flow estimate.")
+            stored_seed = args.seed
+            args.seed = random.randint(0, 2**32 - 1)
```
Please, attach this RNG generation to the very first seed Deforum is being fed, so it won't break reproducibility. See the `s` variable Matisse's last PR introduces

reallybigname replied:
I don't understand. Redo is random during redo. It stores the seed going in and restores the seed going out.  That's the most reproducible it gets. Redo is an inherently non-reproducible process, just like with any random seed.

kabachuha commented on scripts/deforum_helpers/render.py:
```diff
@@ -484,12 +485,26 @@ def render_animation(args, anim_args, video_args, parseq_args, loop_args, contro
+                args.init_sample = Image.fromarray(cv2.cvtColor(disposable_image, cv2.COLOR_BGR2RGB))
+            del(disposable_image)
+            args.seed = stored_seed
 
```
Add python's garbage collection trigger here, so it will clear the deleted variables and not clutter the RAM

reallybigname replied:
I don't know how to do that. I'm deleting the image right there. But I guess that's different than garbage collection. Deleting it was an afterthought. I mean, there are lots of things that don't get deleted during various processes. If there's a way to optimize those, that would be cool. I just thought it would be a good idea to delete those, since they are disoposable.

reallybigname replied:
Probably should be addressed in a special garbage collection PR, so it's not just patching one little thing.

reallybigname committed:
fixed colors drifting during redo and optical flow redo

I just forgot that color would drift during repeated generations. Fixed. Now it's really consistent... and redo is incredible.

reallybigname committed:
made diffusion redo only color match on last one out

After testing it became obvious that part of the magic of the redo was letting the colors drift during redo... so, I just color matched on the way out, and it maintains it's cool factor, while maintaining colors!

reallybigname committed:
Fixed one-off error in Redo

My most embarrassing commit.. I tested a bunch of different numbers of redos... except for 1... which caused an error and made me realize I had a one off problem... Tested 0, 1, 2, 3... working properly now. 1 redo yields 2 generations... as it should.

reallybigname committed:
Removed test code.. whoops.

Sorry, I'm tired... I had added this to test something and forgot to remove, and didn't notice I commited it.

reallybigname committed:
Added optical flow cadence method selection

- Uses the same variable, optical_flow_cadence but now uses "None" for disabled and has the three optical flow methods.
- Improved reporting of cadence frames to include optical flow cadence type if one is selected

kabachuha commented:
@reallybigname I think I'll merge this PR into a temp branch after a second review, and just fix those two little issues myself
